lib/crypto/md5.toit

```diff
@@ -0,0 +1,149 @@
+// Copyright (C) 2023 Toitware ApS. All rights reserved.
+// Use of this source code is governed by an MIT-style license that can be
+// found in the lib/LICENSE file.
+
+import binary show LITTLE_ENDIAN
+import .checksum
+
+class MD5 extends Checksum:
+  static SHIFTS_ ::= [
+    07, 12, 17, 22, 07, 12, 17, 22, 07, 12, 17, 22, 07, 12, 17, 22, 05, 09, 14,
+    20, 05, 09, 14, 20, 05, 09, 14, 20, 05, 09, 14, 20, 04, 11, 16, 23, 04, 11,
+    16, 23, 04, 11, 16, 23, 04, 11, 16, 23, 06, 10, 15, 21, 06, 10, 15, 21, 06,
+    10, 15, 21, 06, 10, 15, 21
+  ]
+
+  static NOISE_ ::= [
+    0xd76aa478, 0xe8c7b756, 0x242070db, 0xc1bdceee, 0xf57c0faf, 0x4787c62a,
+    0xa8304613, 0xfd469501, 0x698098d8, 0x8b44f7af, 0xffff5bb1, 0x895cd7be,
+    0x6b901122, 0xfd987193, 0xa679438e, 0x49b40821, 0xf61e2562, 0xc040b340,
+    0x265e5a51, 0xe9b6c7aa, 0xd62f105d, 0x02441453, 0xd8a1e681, 0xe7d3fbc8,
+    0x21e1cde6, 0xc33707d6, 0xf4d50d87, 0x455a14ed, 0xa9e3e905, 0xfcefa3f8,
+    0x676f02d9, 0x8d2a4c8a, 0xfffa3942, 0x8771f681, 0x6d9d6122, 0xfde5380c,
+    0xa4beea44, 0x4bdecfa9, 0xf6bb4b60, 0xbebfbc70, 0x289b7ec6, 0xeaa127fa,
+    0xd4ef3085, 0x04881d05, 0xd9d4d039, 0xe6db99e5, 0x1fa27cf8, 0xc4ac5665,
+    0xf4292244, 0x432aff97, 0xab9423a7, 0xfc93a039, 0x655b59c3, 0x8f0ccc92,
+    0xffeff47d, 0x85845dd1, 0x6fa87e4f, 0xfe2ce6e0, 0xa3014314, 0x4e0811a1,
+    0xf7537e82, 0xbd3af235, 0x2ad7d2bb, 0xeb86d391
+  ]
+
+  size_/int := 0
+  buffer_/ByteArray? := ByteArray 64
+  digest_/ByteArray? := null
+
+  a_/int := 0x67452301
+  b_/int := 0xefcdab89
+  c_/int := 0x98badcfe
+  d_/int := 0x10325476
+
+  add data from/int to/int -> none:
+    if not buffer_: throw "ALREADY_CLOSED"
+    slice := data[from..to]
+    add_bytes_ (slice is ByteArray ? slice : slice.to_byte_array)
```

Comment:
You can't take arbitrary slices of strings because of malformed UTF-8.

```suggestion
    if data is string:
      add_bytes_ (data.to_byte_array from to)
    else:
      add_bytes_ data[from..to]
```

Reply:
As discussed, we will leave this out.